src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/AuthenticationPolicy.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using Azure.Base.Http;
+using Azure.Base.Http.Pipeline;
+
+namespace Azure.ApplicationModel.Configuration
+{
+    internal class AuthenticationPolicy : HttpPipelinePolicy
+    {
+        private readonly string _credential;
+        private readonly byte[] _secret;
+
+        public AuthenticationPolicy(string credential, byte[] secret)
+        {
+            _credential = credential;
+            _secret = secret;
+        }
+
+        public override async Task ProcessAsync(HttpPipelineMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            string contentHash;
+
+            using (var alg = SHA256.Create())
+            {
+                using (var memoryStream = new MemoryStream())
+                using (var contentHashStream = new CryptoStream(memoryStream, alg, CryptoStreamMode.Write))
+                {
+                    if (message.Request.Content != null)
+                    {
+                        await message.Request.Content.WriteTo(contentHashStream, message.Cancellation);
+                    }
+                }
+
+                contentHash = Convert.ToBase64String(alg.Hash);
+            }
+
+            using (var hmac = new HMACSHA256(_secret))
+            {
+                var uri = message.Request.Uri;
+                var host = uri.Host;
+                var pathAndQuery = uri.PathAndQuery;
+
+                string verb = message.Request.Method.ToString().ToUpper();
```

Comment:
Can we write a reusable and not-allocating routine for this?

Reply:
Yes.

Should we switch the method to be a string? Seems like all we going is converting back and forth from the string.

Reply:
I think enum is better as it prevent misspellings and is supported by intellisense. To avoid the cost of converting, we can have a lookup table.

Reply:
```
public struct HttpVerb
{
    public HttpVerb Get { get; } = new HttpVerb("GET");
    public HttpVerb Post { get; } = new HttpVerb("POST");

    ...

	public HttpVerb(string name)
    {
       ...
    }
}
```

Allows it to be strongly typed while allowing easy string interop and custome verbs too.

Reply:
What's the benefit of this over an enum? I see mainly negatives. Enum has first class intellisense support. This pattern requires people to know about the statics.

Reply:
This looks very similar to enum from intellisense point of view. There are two benefits:

1. Allows for custom methods without a need of changing the enum
2. Simplifies conversion from/to string

Reply:
I don't think we will be using custom method. And then conversion is all internal code. Enum is just a natural type kind for these kinds of things, IMO

Reply:
What type do you want to put enum<->string conversion methods on?

Reply:
That's another benefit of struct, you don't need another type to hold formatting methods and `verb.ToString()` just works instead of having to type `HttpVerbConverter.ToString(verb)`